### PR TITLE
Added Held Slot handler

### PR DESCRIPTION
**Pull request by CiviledCode**
- Removed session.UpdateHeldSlot
- Added player.SetHeldSlot
- Added HeldSlotChange to handler

### CODE REVIEW

File: server/player/handler.go [hunk 1]
```diff
@@ -108,6 +109,11 @@ type Handler interface {
 	// HandleItemPickup handles the player picking up an item from the ground. The item stack laying on the
 	// ground is passed. ctx.Cancel() may be called to prevent the player from picking up the item.
 	HandleItemPickup(ctx *event.Context, i item.Stack)
+
```
Comment (JustTalDevelops) [inline server/player/handler.go:112]: Would remove these empty gaps and the empty comment to match the format.

File: server/player/player.go [hunk 2]
```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
 	return
 }
 
+// SetHeldSlot updates the held slot of the player with the slot passed. It also verifies that the item within the spot
+// is the correct item.
+// This invokes the HandleHeldSlotChange handler.
+func (p *Player) SetHeldSlot(slot int, expected item.Stack) error {
+	if slot > 8 {
```
Comment (T14Raptor) [inline server/player/player.go:375]: Why'd you remove all the documentation for this method?

File: server/player/player.go [hunk 3]
```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
...
+		return nil
+	}
+
+	p.ReleaseItem()
+
+	clientSideItem := expected
+	actual, _ := p.inv.Item(slot)
+
+	if !clientSideItem.Equal(actual) {
+		return fmt.Errorf("client side item isn't equal to expected slot item")
```
Comment (T14Raptor) [inline server/player/player.go:389]: This code no longer belongs here now that it's a player method, should be moved to packet handling (check a0bb8dde26bc089aac89d2625cb45a8a1dfe6523).

File: server/player/player.go [hunk 4]
```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
...
+func (p *Player) SetHeldSlot(slot int, expected item.Stack) error {
+	if slot > 8 {
+		return fmt.Errorf("new held slot exceeds hotbar range 0-8: slot is %v", slot)
+	}
+
+	if p.heldSlot.Swap(uint32(slot)) == uint32(slot) {
+		return nil
+	}
+
+	p.ReleaseItem()
```
Comment (T14Raptor) [inline server/player/player.go:383]: Should just set `(Player).usingItem` to false here.

File: server/player/player.go [hunk 5]
```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
 	return
 }
 
+// SetHeldSlot updates the held slot of the player with the slot passed. It also verifies that the item within the spot
+// is the correct item.
+// This invokes the HandleHeldSlotChange handler.
+func (p *Player) SetHeldSlot(slot int, expected item.Stack) error {
```
Comment (T14Raptor) [inline server/player/player.go:374]: Passing an expected item stack no longer makes sense here, should be removed.

File: server/player/player.go [hunk 6]
```diff
@@ -367,6 +368,37 @@ func (p *Player) Transfer(address string) (err error) {
 	return
 }
 
+// SetHeldSlot updates the held slot of the player with the slot passed. It also verifies that the item within the spot
```
Comment (T14Raptor) [inline server/player/player.go:371]: Should remove the middle sentence in this doc.

**Review by T14Raptor — CHANGES_REQUESTED** [review]

**Comment by JustTalDevelops** [comment]
Any update on this?

**Pull request closed without merging**